Employee attrition is a critical challenge for organizations, as losing skilled employees can lead to increased recruitment costs, reduced productivity, and disruption of team dynamics. Predicting which employees are likely to leave allows HR departments to take proactive measures to improve retention and maintain organizational stability.

This project focuses on analyzing an HR dataset containing employee demographics, job-related attributes, and workplace information to predict employee attrition. The dataset includes both numerical features (such as Age, DistanceFromHome, TotalWorkingYears) and categorical features (such as Department, JobRole, OverTime), which are preprocessed and used as inputs for machine learning models.

The primary objective is to build a predictive model that can accurately classify whether an employee is likely to stay or leave. The insights generated from this analysis can help organizations identify risk factors for attrition and design effective employee engagement and retention strategies.

In [1]:
#importing the libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
import os
os.chdir('/content/drive/MyDrive/Boosting')

Step1: Loading the datasets

In [4]:
df = pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv')

In [5]:
df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [6]:
#Checking the number of 'Yes' and 'No' in 'Attrition'
df['Attrition'].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

In [7]:
#checking any missing value in the data
df.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [8]:
#checking the info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

Step2: Feature Engineering

The numeric and categorical fields need to be treated separately.The following few steps separate the numeric and categorical fields and drops the target field 'Attrition' from the feature set.

In [9]:
# Extracting the numeric values
df_num = df.select_dtypes(include = ['int64'])

In [10]:
df_num.head()

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  EmployeeNumber  \
0   41       1102                 1          2              1               1   
1   49        279                 8          1              1               2   
2   37       1373                 2          2              1               4   
3   33       1392                 3          4              1               5   
4   27        591                 2          1              1               7   

   EnvironmentSatisfaction  HourlyRate  JobInvolvement  JobLevel  ...  \
0                        2          94               3         2  ...   
1                        3          61               2         2  ...   
2                        4          92               2         1  ...   
3                        4          56               3         1  ...   
4                        1          40               3         1  ...   

   RelationshipSatisfaction  StandardHours  StockOptionLevel  \
0                         1             80                 0   
1                         4             80                 1   
2                         2             80                 0   
3                         3             80                 0   
4                         4             80                 1   

   TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
0                  8                      0                1               6   
1                 10                      3                3              10   
2                  7                      3                3               0   
3                  8                      3                3               8   
4                  6                      3                3               2   

   YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                   4                        0                     5  
1                   7                        1                     7  
2                   0                        0                     0  
3                   7                        3                     0  
4                   2                        2                     2  

[5 rows x 26 columns]

In [11]:
# extracting the categorical values
df_cat = df.select_dtypes(include = ['object'])

In [12]:
df_cat.head()

Attrition     BusinessTravel              Department EducationField  Gender  \
0       Yes      Travel_Rarely                   Sales  Life Sciences  Female   
1        No  Travel_Frequently  Research & Development  Life Sciences    Male   
2       Yes      Travel_Rarely  Research & Development          Other    Male   
3        No  Travel_Frequently  Research & Development  Life Sciences  Female   
4        No      Travel_Rarely  Research & Development        Medical    Male   

                 JobRole MaritalStatus Over18 OverTime  
0        Sales Executive        Single      Y      Yes  
1     Research Scientist       Married      Y       No  
2  Laboratory Technician        Single      Y      Yes  
3     Research Scientist       Married      Y      Yes  
4  Laboratory Technician       Married      Y       No

In [13]:
# dropping the Attrition from df_cat

df_cat = df_cat.drop('Attrition', axis =1)

In [14]:
df_cat.head()

BusinessTravel              Department EducationField  Gender  \
0      Travel_Rarely                   Sales  Life Sciences  Female   
1  Travel_Frequently  Research & Development  Life Sciences    Male   
2      Travel_Rarely  Research & Development          Other    Male   
3  Travel_Frequently  Research & Development  Life Sciences  Female   
4      Travel_Rarely  Research & Development        Medical    Male   

                 JobRole MaritalStatus Over18 OverTime  
0        Sales Executive        Single      Y      Yes  
1     Research Scientist       Married      Y       No  
2  Laboratory Technician        Single      Y      Yes  
3     Research Scientist       Married      Y      Yes  
4  Laboratory Technician       Married      Y       No

In [15]:
# encoding using pandas to get dummies

df_cat_encoded = pd.get_dummies(df_cat)

In [16]:
df_cat_encoded.head()

BusinessTravel_Non-Travel  BusinessTravel_Travel_Frequently  \
0                      False                             False   
1                      False                              True   
2                      False                             False   
3                      False                              True   
4                      False                             False   

   BusinessTravel_Travel_Rarely  Department_Human Resources  \
0                          True                       False   
1                         False                       False   
2                          True                       False   
3                         False                       False   
4                          True                       False   

   Department_Research & Development  Department_Sales  \
0                              False              True   
1                               True             False   
2                               True             False   
3                               True             False   
4                               True             False   

   EducationField_Human Resources  EducationField_Life Sciences  \
0                           False                          True   
1                           False                          True   
2                           False                         False   
3                           False                          True   
4                           False                         False   

   EducationField_Marketing  EducationField_Medical  ...  \
0                     False                   False  ...   
1                     False                   False  ...   
2                     False                   False  ...   
3                     False                   False  ...   
4                     False                    True  ...   

   JobRole_Research Director  JobRole_Research Scientist  \
0                      False                       False   
1                      False                        True   
2                      False                       False   
3                      False                        True   
4                      False                       False   

   JobRole_Sales Executive  JobRole_Sales Representative  \
0                     True                         False   
1                    False                         False   
2                    False                         False   
3                    False                         False   
4                    False                         False   

   MaritalStatus_Divorced  MaritalStatus_Married  MaritalStatus_Single  \
0                   False                  False                  True   
1                   False                   True                 False   
2                   False                  False                  True   
3                   False                   True                 False   
4                   False                   True                 False   

   Over18_Y  OverTime_No  OverTime_Yes  
0      True        False          True  
1      True         True         False  
2      True        False          True  
3      True        False          True  
4      True         True         False  

[5 rows x 29 columns]

Scaling the numeric fields

The numeric fields have been scaled next for best results. StandardScaler() has been used for the same. After scaling the numeric features, they will be merged with the categorical features.

In [17]:
# scaling the numeric fields
from sklearn.preprocessing import StandardScaler

In [18]:
scaler = StandardScaler()


In [19]:
df_num_scaled = scaler.fit_transform(df_num)



In [20]:
df_num_scaled

array([[ 0.4463504 ,  0.74252653, -1.01090934, ..., -0.0632959 ,
        -0.67914568,  0.24583399],
       [ 1.32236521, -1.2977746 , -0.14714972, ...,  0.76499762,
        -0.36871529,  0.80654148],
       [ 0.008343  ,  1.41436324, -0.88751511, ..., -1.16768726,
        -0.67914568, -1.15593471],
       ...,
       [-1.08667552, -1.60518328, -0.64072665, ..., -0.61549158,
        -0.67914568, -0.31487349],
       [ 1.32236521,  0.54667746, -0.88751511, ...,  0.48889978,
        -0.67914568,  1.08689522],
       [-0.32016256, -0.43256792, -0.14714972, ..., -0.33939374,
        -0.36871529, -0.59522723]])

In [21]:
#converting num array to data frame
df_num_scaled = pd.DataFrame(df_num_scaled, columns = df_num.columns)

In [22]:
df_num_scaled.head()

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
0  0.446350   0.742527         -1.010909  -0.891688            0.0   
1  1.322365  -1.297775         -0.147150  -1.868426            0.0   
2  0.008343   1.414363         -0.887515  -0.891688            0.0   
3 -0.429664   1.461466         -0.764121   1.061787            0.0   
4 -1.086676  -0.524295         -0.887515  -1.868426            0.0   

   EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
0       -1.701283                -0.660531    1.383138        0.379672   
1       -1.699621                 0.254625   -0.240677       -1.026167   
2       -1.696298                 1.169781    1.284725       -1.026167   
3       -1.694636                 1.169781   -0.486709        0.379672   
4       -1.691313                -1.575686   -1.274014        0.379672   

   JobLevel  ...  RelationshipSatisfaction  StandardHours  StockOptionLevel  \
0 -0.057788  ...                 -1.584178            0.0         -0.932014   
1 -0.057788  ...                  1.191438            0.0          0.241988   
2 -0.961486  ...                 -0.658973            0.0         -0.932014   
3 -0.961486  ...                  0.266233            0.0         -0.932014   
4 -0.961486  ...                  1.191438            0.0          0.241988   

   TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
0          -0.421642              -2.171982        -2.493820       -0.164613   
1          -0.164511               0.155707         0.338096        0.488508   
2          -0.550208               0.155707         0.338096       -1.144294   
3          -0.421642               0.155707         0.338096        0.161947   
4          -0.678774               0.155707         0.338096       -0.817734   

   YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0           -0.063296                -0.679146              0.245834  
1            0.764998                -0.368715              0.806541  
2           -1.167687                -0.679146             -1.155935  
3            0.764998                 0.252146             -1.155935  
4           -0.615492                -0.058285             -0.595227  

[5 rows x 26 columns]

In [23]:
df_num_scaled.shape

(1470, 26)

In [24]:
df_cat_encoded.shape

(1470, 29)

In [25]:
# combining the both num and cat data frame
df_final = pd.concat([df_num_scaled, df_cat_encoded], axis=1)

In [26]:
df_final.head()

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
0  0.446350   0.742527         -1.010909  -0.891688            0.0   
1  1.322365  -1.297775         -0.147150  -1.868426            0.0   
2  0.008343   1.414363         -0.887515  -0.891688            0.0   
3 -0.429664   1.461466         -0.764121   1.061787            0.0   
4 -1.086676  -0.524295         -0.887515  -1.868426            0.0   

   EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
0       -1.701283                -0.660531    1.383138        0.379672   
1       -1.699621                 0.254625   -0.240677       -1.026167   
2       -1.696298                 1.169781    1.284725       -1.026167   
3       -1.694636                 1.169781   -0.486709        0.379672   
4       -1.691313                -1.575686   -1.274014        0.379672   

   JobLevel  ...  JobRole_Research Director  JobRole_Research Scientist  \
0 -0.057788  ...                      False                       False   
1 -0.057788  ...                      False                        True   
2 -0.961486  ...                      False                       False   
3 -0.961486  ...                      False                        True   
4 -0.961486  ...                      False                       False   

   JobRole_Sales Executive  JobRole_Sales Representative  \
0                     True                         False   
1                    False                         False   
2                    False                         False   
3                    False                         False   
4                    False                         False   

   MaritalStatus_Divorced  MaritalStatus_Married  MaritalStatus_Single  \
0                   False                  False                  True   
1                   False                   True                 False   
2                   False                  False                  True   
3                   False                   True                 False   
4                   False                   True                 False   

   Over18_Y  OverTime_No  OverTime_Yes  
0      True        False          True  
1      True         True         False  
2      True        False          True  
3      True        False          True  
4      True         True         False  

[5 rows x 55 columns]

In [27]:
# assigning the feature and target variable
y= df['Attrition']
X= df_final

Train and test split

In [28]:
from sklearn.model_selection import train_test_split

In [29]:
X_train, X_test , y_train, y_test = train_test_split(X, y , train_size = .7, random_state =100)

Step 3: Model Fitting

Adaboost Classifier

The most important parameters are base_estimator, n_estimators and learning_rate.
1. base_estimator - It is the learning algorithm to use to train the weak models. The default Learning Algorithm is DecisionTreeClassifier with Max Depth of 1
2. n_estimators - It is the number of models to iteratively train.

3.learning_rate - It is the contribution of each model to the weights and default value for it is 1.

There is a trade-off between learning_rate and n_estimators. Reducing the learning rate will forcing the model train slower (but sometimes resulting in better performance scores). Decreasing the learning rate L makes the coefficients α_m smaller, which reduces the amplitude of the sample_weights at each step (As per weight formula we use at each step for updating weights).

In [30]:
from sklearn.ensemble import AdaBoostClassifier

In [31]:
# building the model
adaboost = AdaBoostClassifier(n_estimators =200, random_state =100)

In [32]:
# fitting the model
adaboost.fit(X_train, y_train)

AdaBoostClassifier(n_estimators=200, random_state=100)

In [33]:
# predicting the value of y

y_pred = adaboost.predict(X_test)

In [34]:
from sklearn.metrics import accuracy_score, confusion_matrix

In [35]:
accuracy_score(y_test, y_pred)

0.8639455782312925

In [36]:
confusion_matrix(y_test, y_pred)

array([[355,  16],
       [ 44,  26]])

Conclusion

The HR Employee Attrition project aimed to predict whether an employee is likely to leave the organization using historical employee data. The dataset contained both numerical features (like Age, DistanceFromHome, MonthlyIncome) and categorical features (like Department, JobRole, OverTime), which were appropriately preprocessed — numerical features were scaled, and categorical features were one-hot encoded.

The model used for prediction was AdaBoost Classifier, a robust ensemble learning technique that combines multiple weak learners to improve classification performance. After training and testing the model:

The accuracy achieved was 86.4%, indicating that the model can correctly classify employees’ attrition status in the majority of cases.

The confusion matrix revealed that the model predicts employees who stay (non-attrition) very well but has some difficulty predicting actual attrition cases, which is typical in imbalanced datasets.

Overall, the model demonstrates that machine learning can be effectively leveraged to identify employees at risk of leaving, enabling HR departments to implement proactive retention strategies. Further improvements could include addressing class imbalance (e.g., using SMOTE), feature selection, and experimenting with other advanced models like XGBoost or Random Forest for potentially higher predictive performance.